src/shell.ts

```diff
@@ -1,475 +1,150 @@
-import cp from "child_process"
+import { ChildProcess, spawn } from "child_process"
+import { randomUUID } from "crypto"
 import fs from "fs"
 import path from "path"
 import { promisify } from "util"
 
-import { CommandExecutor, CommandOutput, Context, ToString } from "./types"
-import { displayCommand, intoError, redact } from "./utils"
+import { Logger } from "./logger"
+import { Context, ToString } from "./types"
+import { displayCommand, redact } from "./utils"
 
 export const fsExists = promisify(fs.exists)
 export const fsReadFile = promisify(fs.readFile)
 export const fsWriteFile = promisify(fs.writeFile)
-const fsRmdir = promisify(fs.rmdir)
 const fsMkdir = promisify(fs.mkdir)
 const fsUnlink = promisify(fs.unlink)
-const cpExec = promisify(cp.exec)
 
-enum RetryContext {
-  CompilationError = "compilation error",
+export const ensureDir = async (dir: string) => {
+  if (!(await fsExists(dir))) {
+    await fsMkdir(dir, { recursive: true })
```

Comment:
`await fsMkdir(dir, { recursive: true })` won't throw an error on a directory that already exists. 

Using `fs.exists` before any file-system operation is [being discouraged](https://nodejs.org/dist/latest-v16.x/docs/api/fs.html#fsexistspath-callback) by node.js team, because it opens a path to race conditions, and wasted syscalls.

So, I suggest just using `await fsMkdir(dir, { recursive: true })` instead of this function